### PR TITLE
Revived: Map/reduce type variance improvements

### PR DESCRIPTION
**Pull request by mtopolnik**
Revives #4872, which needed rebasing.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/mapreduce/impl/AbstractJob.java [hunk 1]
```diff
@@ -85,7 +86,7 @@ public AbstractJob(String name, JobTracker jobTracker, KeyValueSource<KeyIn, Val
     }
 
     @Override
-    public Job<KeyIn, ValueIn> onKeys(Iterable<KeyIn> keys) {
+    public Job<KeyIn, ValueIn> onKeys(Iterable<? extends KeyIn> keys) {
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/mapreduce/impl/AbstractJob.java]: This method seems to lack code coverage: https://hazelcast-l337.sonar.cloudbees.com/dashboard/index/12063?display_title=true&metric=coverage

File: hazelcast/src/main/java/com/hazelcast/mapreduce/impl/AbstractJob.java [hunk 2]
```diff
@@ -138,7 +139,7 @@ protected void prepareKeyPredicate() {
         }
     }
 
-    private void addKeys(Iterable<KeyIn> keys) {
+    private void addKeys(Iterable<? extends KeyIn> keys) {
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/mapreduce/impl/AbstractJob.java]: No code coverage as well: https://hazelcast-l337.sonar.cloudbees.com/dashboard/index/12063?display_title=true&metric=coverage

**Comment by Donnerbart** [comment]
It's not visible in SonarQube if the interfaces are used (so I had to check that in IntelliJ). `Job` is implemented by `AbstractJob` and `MappingJob` by the inner class `AbstractJob.MappingJobImpl`.

There is a lot of missing code coverage in the `AbstractJob`, especially the inner class `MappingJobImpl`. This might be an JaCoCo issue or real missing code coverage.

Beside that the compiler should tell us if the changes are okay. And since all changed classes are somewhere used in tests we should see a failure in the PR builder.

So :+1: (once green).

**Comment by mtopolnik** [comment]
Even though I am the formal submitter, this is not my code so I should be allowed to vote.

:+1:

**Pull request merged**